### PR TITLE
Enhance the Keycloak Operator with Network Policies

### LINKED ISSUE #34659: [Operator] Enhance the Keycloak Operator with Network Policies
### Description

Extend the Keycloak Operator with an option for adding NetworkPolicies to prevent accesses from unexpected namespace. 

### Discussion

_No response_

### Motivation

_No response_

### Details

Suggested network policy:

* JGroups Port 7800 and 57800 from other Keycloak pods only. Users may change these ports with custom stacks or by setting the property `-Djgroups.bind.address=x`. It will be outside of the scope of this initial change.
* Management port 9000, when metrics are enabled and restricted to the monitoring namespace
* Main port 8843 or 8080 by checking if TLS is enabled or not and restricted to ingress.

### PR DESCRIPTION
**Pull request by pruivo**
Closes #34659

Minikube does not support Network Policies as described here: https://minikube.sigs.k8s.io/docs/handbook/network_policy/

I've added the `calico` addon as suggested by the same webpage. 

```[tasklist]
### Peding  Tasks
- [x] Documentation
- [x] Manual test in OpenShift
```

### CODE REVIEW

**Comment by pruivo** [comment]
Updated
* Fixed the test suite. :crossed_fingers: GHA uses a custom and optimized image that fails to start when changing the build time attributes.

**Comment by pruivo** [comment]
Single site deployments on OpenShift seems to work properly
```
$ oc describe networkpolicies.networking.k8s.io keycloak-network-policy 
Name:         keycloak-network-policy
Namespace:    pruivo-keycloak
Created on:   2024-11-12 12:51:24 +0000 WET
Labels:       app=keycloak
              app.kubernetes.io/instance=keycloak
              app.kubernetes.io/managed-by=keycloak-operator
Annotations:  javaoperatorsdk.io/previous: e92c6f76-7c1d-4041-8975-d61d90ffc324
Spec:
  PodSelector:     app=keycloak,app.kubernetes.io/instance=keycloak,app.kubernetes.io/managed-by=keycloak-operator
  Allowing ingress traffic:
    To Port: 8443/TCP
    From: <any> (traffic not restricted by source)
    ----------
    To Port: 7800/TCP
    To Port: 57800/TCP
    From:
      PodSelector: app=keycloak,app.kubernetes.io/instance=keycloak,app.kubernetes.io/managed-by=keycloak-operator
    ----------
    To Port: 9000/TCP
    From: <any> (traffic not restricted by source)
  Not affecting egress traffic
  Policy Types: Ingress
```

**Comment by pruivo** [comment]
Multi-site deployment confirmed without JGroups ports:
```
$ oc describe networkpolicies.networking.k8s.io keycloak-network-policy 
Name:         keycloak-network-policy
Namespace:    pruivo-keycloak
Created on:   2024-11-12 14:01:10 +0000 WET
Labels:       app=keycloak
              app.kubernetes.io/instance=keycloak
              app.kubernetes.io/managed-by=keycloak-operator
Annotations:  javaoperatorsdk.io/previous: 1f15f9df-8c1e-42db-b304-e11cfbc83ab7
Spec:
  PodSelector:     app=keycloak,app.kubernetes.io/instance=keycloak,app.kubernetes.io/managed-by=keycloak-operator
  Allowing ingress traffic:
    To Port: 8443/TCP
    From: <any> (traffic not restricted by source)
    ----------
    To Port: 9000/TCP
    From: <any> (traffic not restricted by source)
  Not affecting egress traffic
  Policy Types: Ingress
```

**Comment by pruivo** [comment]
The test suite seems more unstable with the network plugin. If it causes too much trouble, I run the test in a different workflow.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you @pruivo - this looks like a great addition to me when reviewing the Network policies you extracted and test tests.

I agree that this would be experimental until we offer also a way to parameterize the "from" part on the ingress and the metrics. This is the only option that would keep me from making it preview and eventually fully supported. 

See below for some nitpicks on the docs. I would be good to have a more in-depth review of the CND team.

**Comment by ahus1** [comment]
@pruivo - can you elaborate on the kind of errors you have seen?

> The test suite seems more unstable with the network plugin. If it causes too much trouble, I run the test in a different workflow.

**Comment by pruivo** [comment]
@ahus1 known issues:

*  DNS_PING takes time to fetch the correct IPs, see my comment: https://github.com/keycloak/keycloak/pull/34788#issuecomment-2468810727
* `ClusteringTest.testKeycloakScaleAsExpected` fails with `calipo` network plugin. Switching to the `cilium` network plugin and increasing the CPU limit stabilized this test. The keycloak container fails to start; the thread dump is below and the run is here: https://github.com/keycloak/keycloak/actions/runs/11783801029/job/32821913709
```
"main" #1 [87] prio=5 os_prio=0 cpu=3899.82ms elapsed=299.79s tid=0x00007fc07c02ba80 nid=87 runnable  [0x00007fc0840ed000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.Net.poll(java.base@21.0.5/Native Method)
	at sun.nio.ch.NioSocketImpl.park(java.base@21.0.5/NioSocketImpl.java:191)
	at sun.nio.ch.NioSocketImpl.park(java.base@21.0.5/NioSocketImpl.java:201)
	at sun.nio.ch.NioSocketImpl.implRead(java.base@21.0.5/NioSocketImpl.java:309)
	at sun.nio.ch.NioSocketImpl.read(java.base@21.0.5/NioSocketImpl.java:346)
	at sun.nio.ch.NioSocketImpl$1.read(java.base@21.0.5/NioSocketImpl.java:796)
	at java.net.Socket$SocketInputStream.read(java.base@21.0.5/Socket.java:1099)
	at org.postgresql.core.VisibleBufferedInputStream.readMore(VisibleBufferedInputStream.java:192)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:159)
	at org.postgresql.core.VisibleBufferedInputStream.ensureBytes(VisibleBufferedInputStream.java:144)
	at org.postgresql.core.VisibleBufferedInputStream.read(VisibleBufferedInputStream.java:76)
	at org.postgresql.core.PGStream.receiveChar(PGStream.java:476)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2174)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:356)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:341)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:317)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:312)
	at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:220)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:470)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:83)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:185)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:153)
	at org.keycloak.connections.jpa.updater.liquibase.lock.CustomLockService.acquireLock(CustomLockService.java:249)
	at org.keycloak.connections.jpa.updater.liquibase.lock.CustomLockService.waitForLock(CustomLockService.java:206)
	at org.keycloak.connections.jpa.updater.liquibase.lock.CustomLockService.waitForLock(CustomLockService.java:196)
	at org.keycloak.connections.jpa.updater.liquibase.lock.LiquibaseDBLockProvider.lambda$waitForLock$0(LiquibaseDBLockProvider.java:107)
	at org.keycloak.connections.jpa.updater.liquibase.lock.LiquibaseDBLockProvider$$Lambda/0x00007fc03bc0dc70.run(Unknown Source)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:102)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:89)
	at org.keycloak.connections.jpa.updater.liquibase.lock.LiquibaseDBLockProvider.lambda$waitForLock$2(LiquibaseDBLockProvider.java:105)
	at org.keycloak.connections.jpa.updater.liquibase.lock.LiquibaseDBLockProvider$$Lambda/0x00007fc03bbd5358.run(Unknown Source)
	at org.keycloak.models.utils.KeycloakModelUtils.suspendJtaTransaction(KeycloakModelUtils.java:1071)
	at org.keycloak.connections.jpa.updater.liquibase.lock.LiquibaseDBLockProvider.waitForLock(LiquibaseDBLockProvider.java:91)
	at org.keycloak.services.resources.KeycloakApplication$1.run(KeycloakApplication.java:99)
	at org.keycloak.models.utils.KeycloakModelUtils.lambda$runJobInTransaction$1(KeycloakModelUtils.java:274)
	at org.keycloak.models.utils.KeycloakModelUtils$$Lambda/0x00007fc03bbd1738.run(Unknown Source)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:393)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:273)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:263)
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:93)
```

**Comment by ryanemerson** [comment]
> I agree that this would be experimental until we offer also a way to parameterize the "from" part on the ingress and the metrics. This is the only option that would keep me from making it preview and eventually fully supported.

Do you have an idea on how we could parameterize the "from"? Given that the `from` is part of an array entry without a dedicated key, this could be a bit unwieldy. Would you just match on the port section of the entry, e.g.:

Keycloak CR:
```
  ingress:
    enabled: true
  networkPolicy:
    enabled: true
    ingress:
     - ports:
        - protocol: TCP
          port: 9000
       from:
        - namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: openshift-user-workload-monitoring
```

Would override the default definition for port 9000.

**Comment by pruivo** [comment]
Keep it simple and use the Kubernetes resources.  I have something like this in mind:
```yaml
networkPolicy:
    enabled: true
    http: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer`
    https: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer`
    management: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer`
```
The  `NetworkPolicyPeer` is the `ipBlock`, `namespaceSelector`, or `podSelector`.

**Comment by ryanemerson** [comment]
Much better idea @pruivo! The only downside I see is that it's assuming that we will never want want to configure egress rules in our NetworkPolicy, to allow the solution to be more extensible in the future we could do:

```yaml
networkPolicy:
    enabled: true
    ingress:
      http: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer`
      https: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer`
      management: -> List of `io.fabric8.kubernetes.api.model.networking.v1.NetworkPolicyPeer
```

We can then add `networkPolicy.egress` in the future if required.

That being said, I don't have a configuration in mind where a default egress rule makes sense.

**Comment by pruivo** [comment]
I didn't think about egress to be fair @ryanemerson. It seems more complicated to make it work. From the top of my head, Keycloak may establish a connection for:

* Database
* JGroups (data + failure detection, for single-site deployments)
* External Infinispan cluster (multi-site deployments)
* Opentelemetry Tracing 
* Other external systems like Kerberos/LDAP
* CSS style/fonts/analytics (???)
* Red Hat Insights 

+1 to your suggestion to allow future expansion.

**Review by ryanemerson — APPROVED** [review]

**Review by ahus1 — DISMISSED** [review]
Thank you for this PR. I like what I see here. At the same time I'd like to get feedback from @vmuzikar on this approach. Not sure how we want to make features experimental in the Operator. At the same time, I'd say once we add the support for network policy details per port as described in https://github.com/keycloak/keycloak/pull/34788#issuecomment-2488215286, it would be fully supported. That is still within reach for KC26.1, so we shouldn't bother much IMHO.

File: operator/src/test/java/org/keycloak/operator/testsuite/utils/K8sUtils.java [hunk 1]
```diff
@@ -113,18 +115,69 @@ public static String inClusterCurl(KubernetesClient k8sclient, String namespace,
         return inClusterCurl(k8sclient, namespace, "--insecure", "-s", "-o", "/dev/null", "-w", "%{http_code}", url);
     }
 
-    public static String inClusterCurl(KubernetesClient k8sclient, String namespace, String... args) {
-        var podName = "curl-pod";
+    public static void assertJGroupsConnection(KubernetesClient k8sclient, String podIp, String namespace, Map<String, String> labels, boolean connects) {
```
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/testsuite/utils/K8sUtils.java:118]: I think this should rather go to `CRAssert`.

File: .github/workflows/operator-ci.yml [hunk 2]
```diff
@@ -73,7 +73,7 @@ jobs:
           kubernetes version: ${{ env.KUBERNETES_VERSION }}
           github token: ${{ secrets.GITHUB_TOKEN }}
           driver: docker
-          start args: --addons=ingress --memory=${{ env.MINIKUBE_MEMORY }}
+          start args: --addons=ingress --memory=${{ env.MINIKUBE_MEMORY }} --cni cilium --cpus=max
```
Comment (vmuzikar) [inline .github/workflows/operator-ci.yml:76]: My only concern here would be if Cilium affects somehow the tests/cluster stability. Maybe we could try and run a few tens of Operator CI runs to be sure it doesn't introduce new flakiness. WDYT?

**Review by vmuzikar — COMMENTED** [review]
@pruivo Very nice job, thanks!

Added a couple of thoughts, mostly just nitpicks.

@shawkins Do you want to give it a look too?

File: .github/workflows/operator-ci.yml (hunk 2 above)
Comment (pruivo) [reply .github/workflows/operator-ci.yml:76]: Is there a trick to do this or do I need to manually click `re-run jobs`? :smile:

**Comment by pruivo** [comment]
Thank you for the feedback @vmuzikar. I've updated based on your suggestions!

File: .github/workflows/operator-ci.yml (hunk 2 above)
Comment (vmuzikar) [reply .github/workflows/operator-ci.yml:76]: Yes, there's a trick. 🙂 See e.g. https://github.com/vmuzikar/keycloak/blob/85bf6a65d1686a6ab208c8c8ba5e3b47bb55755f/.github/workflows/ci.yml#L77-L78

**Review by vmuzikar — APPROVED** [review]
@pruivo Thank you for the updates! Overall LGTM. I'd like to merge by the end of the week but let's give @shawkins the opportunity to review it as well (considering there was thanksgiving last week).

File: .github/workflows/operator-ci.yml (hunk 2 above)
Comment (vmuzikar) [reply .github/workflows/operator-ci.yml:76]: Checking the stability is not a blocker from my side, just something nice to have.

**Review by shawkins — APPROVED** [review]
This looks great @pruivo.

To double-check, my understanding is that this originated with some issue of cross-talk from pods across namespaces. Was it ever understood why that was happening? Our headless discovery service does give a way to access the pods across namespaces, but I'm not clear how that would actually end up happening because the dns name still contains the namespace. With a network policy if the nodes are discovered across namespaces, but can't communicate, will that lead to a lot of jgroups error messages in the logs?

Depending on the root cause, it seems like we could also have the same issue with 2 keycloaks in the same namespace as well.

Another thought - does this relates at all to some of the rolling upgrade work?  For example using an extra label to segment the caches from old/new pods with incompatible versions - or will that not be an issue moving forward?

**Comment by ahus1** [comment]
The original root cause was never fully resolved. Still using network policies is a best practice in a lot of organizations, so Keycloak should use them as well. 

This does not relate to any rolling-upgrade work, which will follow in future PRs. The only connect is that Pedro and the SRE team get an understanding on how the Keycloak Operator works. Separating caches that are incompatible is not on the roadmap  as of today.

**Comment by pruivo** [comment]
@shawkins failures in JGroups discovery do not cause any errors; the pod creates a new cluster with itself. The Keycloak Operator does a perfect job keeping the cluster separated but issues may arise when some organizations use the Operator mixed with some community helm chart. Making a mistake and setting an incorrect headless service with the helm chart deployment is easy. The NetworkPolicy prevents that mistake from causing any harm. 

@vmuzikar I used your suggestion to run the test suite multiple times and got 20 greens: https://github.com/pruivo/keycloak/actions/runs/12146839236

**Comment by vmuzikar** [comment]
> I used your suggestion to run the test suite multiple times and got 20 greens

Thank you, that looks promising! :)

**Pull request merged**